src/device/net/dev_raw.rs

```diff
@@ -0,0 +1,302 @@
+    /// the reception to complete.
+    ///
+    /// It will submit request to the VirtIO net device and return a token
+    /// identifying the position of the first descriptor in the chain. If there
+    /// are not enough descriptors to allocate, then it returns
+    /// [`Error::QueueFull`].
+    ///
+    /// The caller can then call [`poll_receive`] with the returned token to
+    /// check whether the device has finished handling the request. Once it has,
+    /// the caller must call [`receive_complete`] with the same buffers before
+    /// reading the response.
+    ///
+    /// # Safety
+    ///
+    /// `rx_buf` is still borrowed by the underlying VirtIO net device even after
+    /// this method returns. Thus, it is the caller's responsibility to guarantee
+    /// that they are not accessed before the request is completed in order to
+    /// avoid data races.
+    ///
+    /// [`poll_receive`]: Self::poll_receive
+    /// [`receive_complete`]: Self::receive_complete
+    pub unsafe fn receive_begin(&mut self, rx_buf: &mut [u8]) -> Result<u16> {
+        Self::check_rx_buf_len(rx_buf)?;
+        let token = self.recv_queue.add(&[], &mut [rx_buf])?;
+        if self.recv_queue.should_notify() {
+            self.transport.notify(QUEUE_RECEIVE);
+        }
+        Ok(token)
+    }
+
+    /// Fetches the token of the next completed reception request from the
+    /// used ring and returns it, without removing it from the used ring. If
+    /// there are no pending completed requests it returns [`None`].
+    pub fn poll_receive(&mut self) -> Option<u16> {
+        self.recv_queue.peek_used()
+    }
+
+    /// Completes a transmission operation which was started by [`receive_begin`].
+    ///
+    /// After completion, the `rx_buf` will contain a header followed by the
+    /// received packet. It returns the length of the header and the length of
+    /// the packet.
+    ///
+    /// # Safety
+    ///
+    /// The same buffers must be passed in again as were passed to
+    /// [`receive_begin`] when it returned the token.
+    ///
+    /// [`receive_begin`]: Self::receive_begin
+    pub unsafe fn receive_complete(
+        &mut self,
+        token: u16,
+        rx_buf: &mut [u8],
+    ) -> Result<(usize, usize)> {
+        let len = self.recv_queue.pop_used(token, &[], &mut [rx_buf])? as usize;
+        let packet_len = len.checked_sub(NET_HDR_SIZE).ok_or(Error::IoError)?;
+        Ok((NET_HDR_SIZE, packet_len))
+    }
+
+    /// Transmits a packet to the network, and blocks until the request
+    /// completed. Returns number of bytes transmitted.
+    ///
+    /// The caller needs to fill the `tx_buf` with a header by calling
+    /// [`fill_buffer_header`] before transmission.
+    ///
+    /// [`fill_buffer_header`]: Self::fill_buffer_header
+    pub fn transmit_wait(&mut self, tx_buf: &[u8]) -> Result<usize> {
+        let token = unsafe { self.transmit_begin(tx_buf)? };
```

Comment:
Please add safety comments for the unsafe blocks, or better, use `add_notify_wait_pop` to avoid them entirely.